### ISSUE TITLE
exception with from __future__ annotations

### ISSUE THREAD

**Issue by qcrist** [lovasoa/marshmallow_dataclass#13]
example:
```python
from __future__ import annotations

import marshmallow_dataclass

@marshmallow_dataclass.dataclass
class C:
    s: str
```

result
```python
AttributeError: 'str' object has no attribute '__dataclass_fields__'
```

I believe the problem is in _ _ init _ _.py:
```python
    # Base types
    if typ in _native_to_marshmallow:
        return _native_to_marshmallow[typ](**metadata)
```

since type is 'str' instead of str, the base type isn't found

```
marshmallow-dataclass==6.0.0c1
python==3.7.2
```

**Comment by lovasoa** [comment]
Hello, and thank you for the bug report. It looks like there are several issues around dataclasses and delayed type annotations at the moment. I don't know when/whether this can be done. This is not at trivial as changing the dict you point to.

**Comment by asmodehn** [comment]
Hi,

I am also hitting the same issue.
So I tried the delayed_annotations branch (https://github.com/lovasoa/marshmallow_dataclass/commit/c833a939e8015f53799edc5fe4ebacb6598995ad).

The following code breaks : 
```py
from __future__ import annotations
from dataclasses import field

from marshmallow_dataclass import (
    dataclass,
)  # Importing from marshmallow_dataclass instead of dataclasses


@dataclass(frozen=True)
class Building:

    name: str = field(default="anonymous")

```

with : 
```
Traceback (most recent call last):
  File "/home/alexv/.local/share/JetBrains/Toolbox/apps/PyCharm-C/ch-0/191.6605.12/helpers/pydev/pydevd.py", line 1741, in <module>
    main()
  File "/home/alexv/.local/share/JetBrains/Toolbox/apps/PyCharm-C/ch-0/191.6605.12/helpers/pydev/pydevd.py", line 1735, in main
    globals = debugger.run(setup['file'], None, None, is_module)
  File "/home/alexv/.local/share/JetBrains/Toolbox/apps/PyCharm-C/ch-0/191.6605.12/helpers/pydev/pydevd.py", line 1135, in run
    pydev_imports.execfile(file, globals, locals)  # execute the script
  File "/home/alexv/.local/share/JetBrains/Toolbox/apps/PyCharm-C/ch-0/191.6605.12/helpers/pydev/_pydev_imps/_pydev_execfile.py", line 18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc)
  File "/home/alexv/Projects/marshmallow_dataclass/test_dc.py", line 9, in <module>
    @dataclass(frozen=True)
  File "/home/alexv/Projects/marshmallow_dataclass/marshmallow_dataclass/__init__.py", line 85, in <lambda>
    return add_schema(dc) if _cls else lambda cls: add_schema(dc(cls))
  File "/home/alexv/Projects/marshmallow_dataclass/marshmallow_dataclass/__init__.py", line 101, in add_schema
    clazz.Schema = class_schema(clazz)
  File "/home/alexv/Projects/marshmallow_dataclass/marshmallow_dataclass/__init__.py", line 224, in class_schema
    for field in fields if field.init
  File "/home/alexv/Projects/marshmallow_dataclass/marshmallow_dataclass/__init__.py", line 224, in <genexpr>
    for field in fields if field.init
  File "/home/alexv/Projects/marshmallow_dataclass/marshmallow_dataclass/__init__.py", line 293, in field_for_schema
    return _native_to_marshmallow[typ](**metadata)
  File "/home/alexv/.local/share/virtualenvs/crypy-PletnYfd/lib/python3.7/site-packages/marshmallow/fields.py", line 175, in __init__
    raise ValueError("'missing' must not be set for required fields.")
ValueError: 'missing' must not be set for required fields.

```

Hope this can help for the next step.

More generally, is there any strategy to follow in order to get delayed annotations support ? It is something I am also interested in, and I could try to help.

**Comment by lovasoa** [comment]
@asmodehn Contributions would be welcome. Because of the current ecosystem split, the feature would currently have to be implemented twice: for marshmallow 2 and 3. I would personally prefer waiting for marshmallow 3 to be stable before implementing significant changes in this library, but if you are ready to do it, you are welcome.

The only significant challenge to solve is handling reference cycles. Currently, forward references are stored as strings, that "break" the potential reference cycle. With future annotations, you can very well have :

```py
class X:
   c: X
``` 

and you have to be careful not to enter an infinite loop in this case.

**Comment by ADR-007** [comment]
I made draft PR for that https://github.com/lovasoa/marshmallow_dataclass/pull/54

**Comment by ADR-007** [comment]
I updated the PR 
@lovasoa , could you please review it?

The only thing that you have to make some tricks to use it in a local namespace (you should call MyData.Schema to force schema evaluation). **BUT** usually you should not do it at all. Even more, the marshmallow uses **global** class registry which doesn't resolve name collisions.

I would like to add some check for this like:
```py
    
    def check_schema_name(clazz):
        schema = None
        with suppress(class_registry.RegistryError):
            schema = class_registry.get_class(clazz.__name__)
    
        with suppress(class_registry.RegistryError):
            class_registry.get_class(clazz.__qualname__)
    
        if schema:
            raise NameError(f'The schema name is already used by "{schema!r}"!')
```

but it breaks tests, so let's do it later :)

**Comment by JamieMcKernanKaizen** [comment]
Has there been any update on this? It still seems to be a bug that exists.

**Comment by AleksanderPawlak** [comment]
@lovasoa I've prepared improved version of this PR: https://github.com/lovasoa/marshmallow_dataclass/pull/54 here: https://github.com/lovasoa/marshmallow_dataclass/pull/170 All todos/fixme are resolved, could you please look at it?

**Comment by lovasoa** [comment]
fixed by #186
